File: leetcode/150.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
  int evalRPN(vector<string> &tokens) {
    stack<int> cache;
    for (string &s : tokens) {
      if (s == "+" || s == "-" || s == "*" || s == "/") {
        int v1 = cache.top();
        cache.pop();
        int v2 = cache.top();
        cache.pop();
        if (s == "+") {
          v2 += v1;
        } else if (s == "-") {
          v2 -= v1;
        } else if (s == "*") {
          v2 *= v1;
        } else if (s == "/") {
          v2 /= v1;
        }
        cache.push(v2);
      } else {
        cache.push(atoi(s.c_str()));
      }
    }
    return cache.top();
  }
};

File: leetcode/150.py
from typing import List


class Solution:

    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token == '+':
                stack[-2] += stack[-1]
                stack.pop()
            elif token == '-':
                stack[-2] -= stack[-1]
                stack.pop()
            elif token == '*':
                stack[-2] *= stack[-1]
                stack.pop()
            elif token == '/':
                stack[-2] = int(stack[-2] / stack[-1])
                stack.pop()
            else:
                stack.append(int(token))
        return stack[0]


if __name__ == "__main__":
    Solution().evalRPN(
        ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"])
